### ISSUE TITLE
Case sensitive `find`

### ISSUE THREAD

**Issue by mikermcneil** [balderdashy/waterline#239]
```javascript
User.find({
  role: 'admin',
  name: { 
    equals: 'Gabe'
    caseSensitive: true
  }
})
```

**Comment by pr1ntr** [comment]
afaik, you can use BINARY in mysql to get a case sensitive query.

**Comment by eabovsky** [comment]
+1 for being able to specify case sensitivity on the attribute

**Comment by dmarcelino** [comment]
:+1:

`0.11` could be a good opportunity for this since it will require API/adapter changes (I believe).

**Comment by dmarcelino** [comment]
If everyone agrees with @mikermcneil's suggestion:

> ```javascript
> User.find({
>   role: 'admin',
>   name: { 
>     equals: 'Gabe'
>     caseSensitive: true
>   }
> })
> ```

I don't think we need any waterline core but only adapter and waterline-adapter tests changes. Is this a spec everyone agrees on?

**Comment by winrid** [comment]
I don't agree. It should be case sensitive by default because that is how Mongo behaves.
Trying to make an ORM function entirely the same way for different storage mechanisms simply won't work. There will always be inconsistencies and trying to hide them will only cause confusion.

**Comment by dmarcelino** [comment]
@winrid, as explained by @particlebanana in https://github.com/balderdashy/waterline/issues/239#issuecomment-55673429 the case insensitive default is a legacy issue and one I doubt it will go away. I was hoping we could be constructive and figure a way to improve this in the future without breaking everyone's applications. @mikermcneil's suggestion seems a good one for query level case sensitivity changes.

sails-postgresql already allows changing the default case sensitivity through [configuration](https://github.com/balderdashy/sails-postgresql#configuration). A similar config has been discussed for sails-mongo in https://github.com/balderdashy/sails-mongo/pull/260#issuecomment-95887644 and everybody on the thread agreed upon it.

Summarising:

1. @mikermcneil's suggestion would change case sensitivity at the query level (which can be useful);
2. Through each adapters' configuration we can change the default case sensitivity of each adapter.

**Comment by winrid** [comment]
Sometimes you have to release breaking changes for the sake of good development practices. This would be one of them.

Anyways the adapter level change would be ideal for me.

**Comment by Jolo510** [comment]
@dmarcelino 
+1

**Comment by jfaylon** [comment]
any updates on this?

**Comment by Jolo510** [comment]
I have resulted to use using .native() to write mongodb queries. Seems like a lot of the others have done the same. For SQL queries, I the method used is called .query(). 

http://sailsjs.org/documentation/reference/waterline-orm/models/native

http://sailsjs.org/documentation/reference/waterline-orm/models/query

**Comment by kai23** [comment]
That's what we are using too ( `.query()` )

**Comment by winrid** [comment]
This bypasses the ORM level validations and defeats the purpose of using waterline.

**Comment by Jolo510** [comment]
That is true.

**Comment by haavardw** [comment]
Case insensitive came as a big surprise to me. I ran into problems when looking up a user by name implementing login functionality. When creating users it passes the unique constraint. When searching it all falls apart.

**Comment by Ehesp** [comment]
Also causing problems here, we're trying to do an exact match on a string and it's doing a regexp... taking 200ms for a super simple (or should be) search...

```
"query" : {
        "device_id" : /^5ad2771a\\-35c7\\-4010\\-8420\\-4b8b378050df$/i, 
        "role" : NumberInt(0)
    }, 
```

**Comment by sailsbot** [comment]
Thanks for posting, @mikermcneil.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
@mikermcneil @dmarcelino probably oughtta stay open, at least until it's PRed into the roadmap?  Some serious labels on this one, and it's in a milestone.

**Comment by dmarcelino** [comment]
:+1: for reopening. This is a big one and lots of people have shown interest in this being done.

**Comment by tjwebb** [comment]
+1

**Comment by sailsbot** [comment]
Thanks for posting, @mikermcneil.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by chenweiyj** [comment]
Since I am using utf8 and my data has full-angle data, how can I set the collate for search on a field to do the same thing as following?

`SELECT * FROM tablename where name COLLATE utf8_unicode_ci like '%（商品名：舒澈）%' ;`

**Comment by FewKinG** [comment]
Ran into this issue too, started out pretty well, but not knowing about this issue, as our database slowly grew bigger, some queries got slower and slower. So we got suspicious, I investigated and found out, that quite a few of our indices didn't do any good because queries were transformed to regular expressions. When I found out, that this is expected behavior and the issue known for so long, I couldn't believe it at first. This should be made far more obvious and possible to circumvent by configuration if not by default.

**Comment by neoadventist** [comment]
Yeah this is a big problem for me, compose was complaining about this and i'm going to try to use the native implementation.

**Comment by particlebanana** [comment]
@FewKinG @neoadventist depending on the adapter being used you can toggle the case-sensitivity on/off using the proper flag in your connection configuration. So for example in postgres you would use:

```javascript
postgresql: {
  url: 'postgres://username:password@hostname:port/database',
  wlNext: {
    caseSensitive: true
  }
}
```

It's being [worked on](https://github.com/balderdashy/sails-mysql/pull/304) for MySQL but should work in the Mongo adapter.

Case sensitivity on the query by query level isn't planned for the near future and would need to go through the proposal process.

**Comment by neoadventist** [comment]
thanks @particlebanana, can anyone confirm that this works for the monogo adaptor?

**Comment by neoadventist** [comment]
already implemented the `.native` but interested in the workaround.

**Comment by wkhu** [comment]
+1

**Comment by Salakar** [comment]
Can confirm @particlebanana 's solution works on the mongo adapter. 👍

**Comment by theoinsight** [comment]
I'm having the same issue with mongo doing a regexp. Did anyone find a fix for this?

**Comment by albertpeiro** [comment]
@Salakar what version of sails were you using?
With Sails 1.0 and this config:
```
envMongodbServer: {
      adapter   : 'sails-mongo',
      host      : 'localhost',
      schema: true,
      wlNext: {
        caseSensitive: true
      }
    },
```
I get this:
```
error: Failed to lift app: Error: Invalid configuration for datastore `envMongodbServer`:  Unrecognized options (`wlNext`) specified as config overrides.
This adapter expects only whitelisted properties.
--
See http://sailsjs.com/config/datastores#?the-connection-url for info,
```

**Comment by luislobo** [comment]
@albertpeiro With the latest Waterline, all queries are Case Insensitive. Check the last paragraph in https://github.com/balderdashy/sails-docs/blob/552f6c07da17a2569bcac947e3e5a864fcc8dc81/concepts/ORM/Querylanguage.md

**Comment by anterodev** [comment]
I cant't run a insensitive query with orm using sails-mongo, need any configuration or option?

"sails": "^1.0.0-46",
"sails-hook-orm": "^2.0.0-23",
"sails-mongo": "^1.0.0-11",

**Comment by winrid** [comment]
Waterline cannot be recommended for production environments due to design decisions such as this. Telling users to bypass the ORM to make use of indexes is insane.

**Comment by devinivy** [comment]
Waterline is free, modifiable software, and like any tool it's important to understand its limitations upon adopting it.  The reason this is hard is largely due to the fact that waterline is adapter-based, which is also the ORM's defining feature.  While I don't use waterline anymore and haven't contributed to it in ages, I think it's important that someone express that calling OSS collaborators "insane" for their design decisions, support, or documentation-writing is not cool and definitely not constructive.  So here I am expressing it!  A lot of people avoid collaborating on OSS due to comments like this.  If there's some idea about how to make it easier to implement a feature like this across adapters, that would be very useful and interesting!

**Comment by winrid** [comment]
Provided constructive criticism years ago on the same issue :) It is insane.

> The reason this is hard is largely due to the fact that waterline is adapter-based

No, the reason this is hard is because it is the wrong thing to do. What needs to happen is that it needs to be accepted that you should not be designing a layer of abstraction that functions exactly the same no matter what data store is behind the scenes. Having the interface the same is great - that's the point of an ORM - but for example enforcing what you think is best instead of the DB creators in terms of configuration is not good.

**Comment by devinivy** [comment]
I'm not so interested in the technical aspect here, and I'm not saying you're wrong– the point is that it's more important to be good to each other and respectful of all the free effort that has gone into this tool than to be right.

**Comment by luislobo** [comment]
@winrid this is OS, and MIT, feel free to send your PR! :)
